src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt

```diff
@@ -65,9 +67,65 @@ class PsiElementAssociations(val typeMapping: KotlinTypeMapping, val file: FirFi
                 depth--
             }
         }.visitFile(file, elementMap)
+
+        object : FirDefaultVisitor<Unit, MutableMap<PsiElement, MutableList<FirInfo>>>() {
+            override fun visitElement(element: FirElement, data: MutableMap<PsiElement, MutableList<FirInfo>>) {
+                element.acceptChildren(this, data)
+                if (element is FirResolvedTypeRef) {
+                    // Do not visit FirUserTypeRef, since it's not mappable to a type.
+                    if (element.delegatedTypeRef != null && element.delegatedTypeRef !is FirUserTypeRef) {
+                        // not sure why this isn't taken care of by `FirResolvedTypeRefImpl#acceptChildren()`
+                        element.delegatedTypeRef?.accept(this, data)
+                    }
+                }
+            }
+
+            override fun visitResolvedTypeRef(
+                resolvedTypeRef: FirResolvedTypeRef,
+                data: MutableMap<PsiElement, MutableList<FirInfo>>
+            ) {
+                super.visitResolvedTypeRef(resolvedTypeRef, data)
+                if (resolvedTypeRef.psi is KtTypeReference) {
+                    if (resolvedTypeRef.type is ConeClassLikeType) {
+                        if (resolvedTypeRef.type.typeArguments.isNotEmpty() && resolvedTypeRef.psi is KtTypeReference) {
+                            visitType(resolvedTypeRef.type, resolvedTypeRef.psi as KtTypeReference)
+                        }
+                    } else if (resolvedTypeRef.type is ConeTypeParameterType) {
+                        visitType(resolvedTypeRef.type, resolvedTypeRef.psi as KtTypeReference)
+                    }
+                }
+            }
+
+            private fun visitType(firType: ConeTypeProjection, psiType: KtTypeReference) {
+                if (firType is ConeClassLikeType) {
+                    val psiTypeArguments = psiType.typeElement!!.typeArgumentsAsTypes
+                    if (psiTypeArguments.size != firType.typeArguments.size) {
+                        return
+                    }
+
+                    for ((index, typeArgument) in firType.typeArguments.withIndex()) {
+                        val psiTypeArgument = psiTypeArguments[index] ?: continue
```

Comment:
Wildcards like `*` in `List<*>` will not resolve to a PSI here, so the identifier for `*` will be missing a type.
I don't think it's an issue, because `J.Wildcard` implies the type regardless of the presence of a `JavaType$GenericTypeVariable`.

Reply:
Having this as a comment in the code would be helpful I think.